reana_server/factory.py

```diff
@@ -8,72 +8,4 @@
 
 """Flask-application factory for Reana-Server."""
 
-import logging
-
-from flask import Flask, current_app
-from flask_babelex import Babel
-from flask_menu import Menu as FlaskMenu
-from flask_oauthlib.client import OAuth as FlaskOAuth
-from invenio_accounts import InvenioAccounts
-from invenio_accounts.views import blueprint as blueprint_user
-from invenio_db import InvenioDB
-from invenio_oauthclient import InvenioOAuthClient
-from invenio_oauthclient.views.client import blueprint as blueprint_client
-from invenio_oauthclient.views.settings import blueprint as blueprint_settings
-from reana_commons.config import REANA_LOG_FORMAT, REANA_LOG_LEVEL
-from reana_db.database import Session
-
-
-def create_app(config_mapping=None):
-    """REANA Server application factory."""
-    logging.basicConfig(level=REANA_LOG_LEVEL, format=REANA_LOG_FORMAT, force=True)
-    app = Flask(__name__)
-    app.config.from_object("reana_server.config")
-    if config_mapping:
-        app.config.from_mapping(config_mapping)
-    app.secret_key = "hyper secret key"
-
-    app.session = Session
-
-    Babel(app)
-    FlaskMenu(app)
-    InvenioDB(app)
-    InvenioAccounts(app)
-    FlaskOAuth(app)
-    InvenioOAuthClient(app)
-
-    # Register Invenio OAuth endpoints
-    app.register_blueprint(blueprint_user)
-    app.register_blueprint(blueprint_client)
-    app.register_blueprint(blueprint_settings)
-
-    # Register API routes
-    from .rest import (
-        config,
-        gitlab,
-        ping,
-        secrets,
-        status,
-        users,
-        workflows,
-        info,
-        launch,
-    )  # noqa
-
-    app.register_blueprint(ping.blueprint, url_prefix="/api")
-    app.register_blueprint(workflows.blueprint, url_prefix="/api")
-    app.register_blueprint(users.blueprint, url_prefix="/api")
-    app.register_blueprint(secrets.blueprint, url_prefix="/api")
-    app.register_blueprint(gitlab.blueprint, url_prefix="/api")
-    app.register_blueprint(config.blueprint, url_prefix="/api")
-    app.register_blueprint(status.blueprint, url_prefix="/api")
-    app.register_blueprint(info.blueprint, url_prefix="/api")
-    app.register_blueprint(launch.blueprint, url_prefix="/api")
-
-    @app.teardown_appcontext
-    def shutdown_session(response_or_exc):
-        """Close session on app teardown."""
-        current_app.session.remove()
-        return response_or_exc
-
-    return app
+from invenio_app.factory import create_app
```

Comment:
TODO: find out whether using invenio's `create_app` has unwanted side-effects, but as far as I can tell this is only used in tests

Reply:
This change breaks the generation of the OpenAPI specification file.

Also let's check which invenio endpoints are exposed, even more so after adding `invenio-access` to the dependencies of REANA. Let's also double check that `invenio-access` is really needed

Reply:
REANA is technically a Flask extension (and not a Flask app), as the main Flask app is created by Invenio. This is how `ext.py`/`factory.py` worked before this PR set:

- When running in debug mode, `invenio run ...` is invoked. This calls `invenio_app.factory.create_app`.
- When running in production mode, `uwsgi` is used, and the module configured is `invenio_app.wsgi:application`. This calls `invenio_app.factory.create_app`.
- When running the tests, `reana_server.factory.create_app` is called. This a manually created Flask app with _some_ of Invenio modules loaded.
- When running `generate_openapi_spec.py`, the app is created with `reana_server.factory.create_app`.

Reply:
Keeping the previous create_app for tests and for the `generate_openapi_spec.py` script. Added some comments to clarify the difference between REANA's create_app and Invenio's create_app.

Reply:
I forgot the case of `flas reana-admin` commands. To also make this work, I have made the following changes:
- The app in `app.py` is created with `invenio_app.factory.create_app`. This is the app that Flask autodetects, so we probably want this to be the same as the production app
- The factory method `create_app` in `factory.py` has been renamed to `create_minimal_app`, and it is now used in the tests and in `generate_openapi_spec.py`

Reply:
I have also checked the endpoints exposed by Invenio before/after the upgrade:
- For `/api` nothing changes
- For `/oauth` nothing changes
- There are additional endpoints coming from `invenio-access`, but these are not exposed as the ingress only forwards `/api` and `/oauth` requrests to reana-server ([source](https://github.com/reanahub/reana/blob/8762855d33f859215233439496fc1934532482c4/helm/reana/templates/ingress.yaml#L41-L66))